packages/transport/src/transports/abstract.ts

```diff
@@ -0,0 +1,399 @@
+    /**
+     * todo: document
+     */
+    protected listenPromise?: Deferred<string>;
+
+    /**
+     * todo: document
+     */
+    protected releasePromise?: Deferred<any>;
+    /**
+     * todo: document
+     */
+    protected releasing?: string; // session
+    /**
+     * todo: document
+     */
+    protected abortController: AbortController;
+    /**
+     * todo: document
+     */
+    protected logger: any;
+    constructor({ messages }: ConstructorParams) {
+        super();
+        this.descriptors = [];
+        this.messages = protobuf.Root.fromJSON(messages as protobuf.INamespace);
+        this.name = 'AbstractTransport';
+
+        // @ts-expect-error
+        // I would like to use native AbortController (which should be available also in node)
+        // but tests in trezor-connect are somehow misconfigured and they need this fallback
+        this.abortController = getAbortController();
+
+        // some default logger. it should be possible to set abstract logger from outside
+        this.logger = {
+            debug: (...args: any) => {
+                console.log(`[${this.name}]: `, ...args);
+            },
+            log: (...args: any) => {
+                console.log(`[${this.name}]: `, ...args);
+            },
+        };
+    }
+
+    /**
+     * Tries to initiate transport. Transport might not be available e.g. bridge not running.
+     */
+    abstract init(): AsyncResultWithTypedError<
+        undefined,
+        // webusb only
+        | typeof TRANSPORT_ERRORS.SESSIONS_BACKGROUND_NOT_AVAILABLE
+        // bridge only
+        | typeof TRANSPORT_ERRORS.WRONG_RESULT_TYPE
+        | typeof TRANSPORT_ERRORS.HTTP_ERROR
+        // bridge + webusb
+        | typeof COMMON_ERRORS.UNEXPECTED_ERROR
+    >;
+
+    /**
+     * Setup listeners for device changes (connect, disconnect, change?).
+     * What should it do? Will start emitting DEVICE events after this is fired?
+     * - should call onDescriptorsUpdated in the end
+     */
+    abstract listen(): ResultWithTypedError<undefined, typeof TRANSPORT_ERRORS.ALREADY_LISTENING>;
+
+    /**
+     * List Trezor devices
+     */
+    // todo: type error
+    abstract enumerate(): AsyncResultWithTypedError<Descriptor[], string>;
+
+    /**
+     * Acquire session
+     */
+    abstract acquire({
+        input,
+        first,
+    }: {
+        input: AcquireInput;
+        first?: boolean;
+    }): AsyncResultWithTypedError<
+        string,
+        // webusb
+        | typeof INTERFACE_ERRORS.UNABLE_TO_CLAIM_INTERFACE
+        // bridge
+        | typeof TRANSPORT_ERRORS.DEVICE_DISCONNECTED
+        | typeof TRANSPORT_ERRORS.DEVICE_CLOSED
+        | typeof TRANSPORT_ERRORS.WRONG_RESULT_TYPE // local
+        | typeof TRANSPORT_ERRORS.HTTP_ERROR
+        // webusb + bridge
+        | typeof INTERFACE_ERRORS.DEVICE_NOT_FOUND
+        | typeof SESSION_ERRORS.WRONG_PREVIOUS_SESSION
+        | typeof COMMON_ERRORS.UNEXPECTED_ERROR
+    >;
```

Comment:
I like having all those errors type here. Nice!

Reply:
Cool, actually my plan is to change it a bit in the end. Right now, public methods return either `success` or `error` or `unknownError`. The `unknownError` is useful for development - to find out about all the different errors that might appear. But it doesn't make that much sense to match error messages and either return `error` or `unexpected error` because for example these strings might differ based on platform. 
I am going to use simply `error` more often and return my own errors based on context where it is used. Example:

1. interface.write threw an error.
2. I caught it but I don't know what it is.
3. optional step I am thinking about this one. lets say the caught error.message was either "device disconnected" or "data transfer". I could use `unknownError` to try to guess and return one or another strongly typed error. 
4. But of course I know about context (I was about to write right), so if I return only strongly typed 'data-transfer-error' it should be probably enough.